### PR TITLE
Add Future interface to BigQuery jobs

### LINKED ISSUE #3556: Bigquery Jobs should match the Futures interface
Presently, bigquery uses a custom [`Job`](https://github.com/GoogleCloudPlatform/google-cloud-python/blob/master/bigquery/google/cloud/bigquery/job.py#L85) class (and subclasses) to deal with long-running operations such as queries and data import.

The Job class causes the proliferation of code like this around all of our samples and user code:

```python
def wait_for_job(job):
    while True:
        job.reload()  # Refreshes the state via a GET request.
        if job.state == 'DONE':
            if job.error_result:
                raise RuntimeError(job.errors)
            return
        time.sleep(1)
```

Which in terms means our samples require a level on indirection to accomplish simple tasks:

```python
client = bigquery.Client()
query_job = client.run_async_query(str(uuid.uuid4()), query)
query_job.use_legacy_sql = False
query_job.begin()

wait_for_job(query_job)

rows = query_job.results().fetch_data(max_results=10)
for row in rows:
    print(row)
```

However, our gax-based clients share a common "Long Running Operation" strategy that surfaces these types of on-going operations as a class that implements the [`concurrent.futures.Future`](https://docs.python.org/3/library/concurrent.futures.html#future-objects) interface. This is implemented as [`gax._OperationFuture`](https://github.com/googleapis/gax-python/blob/3d48d1fe3cb149f875328630267ce148565cbb35/google/gax/__init__.py#L549). bigquery unfortunately can not use this as it's an http-only API and gax can only currently be used for gRPC APIs.

We should make the `_BaseJob` class and its subclasses closely conform to the `Future` interface so that the usage is similar to other API client and so that we can simplify usage to:

```python
client = bigquery.Client()
query_job = client.run_async_query(str(uuid.uuid4()), query)
query_job.use_legacy_sql = False
query_job.begin()

# Wait for the job to finish
query_result = query.result()

rows = query_result.fetch_data(max_results=10)
for row in rows:
    print(row)

```

### PR DESCRIPTION
**Pull request by jonparrott**
Resolves #3556 
Towards #3617

### CODE REVIEW

**Comment by jonparrott** [comment]
@tswast this is the usage Im going for: https://gist.github.com/jonparrott/cb0ec50b8c70dfb32693fae021659baa

**Comment by tswast** [comment]
I don't like

```
query_job.result()
rows = query_job.results().fetch_data(max_results=10)
```

I'd prefer if it was

```
rows = query_job.result()
```

Also, the query job needs to be calling [getQueryResults](https://cloud.google.com/bigquery/docs/reference/rest/v2/jobs/getQueryResults) and checking for `jobComplete` in the response rather than polling the job state. The reason for this is that [getQueryResults](https://cloud.google.com/bigquery/docs/reference/rest/v2/jobs/getQueryResults) will wait for the job to complete or timeout (whichever comes first), whereas [get Job](https://cloud.google.com/bigquery/docs/reference/rest/v2/jobs/get) returns immediately.

File: core/google/cloud/future/base.py [hunk 1]
```diff
@@ -89,15 +93,52 @@ def __init__(self):
...
         """
-        # pylint: disable=missing-raises
-        raise NotImplementedError()
+        if self._result_set:
+            return
+
+        retry_on = tenacity.retry_if_result(
+            functools.partial(operator.is_not, True))
+        # Use exponential backoff with jitter.
+        wait_on = (
```
Comment (tswast) [inline core/google/cloud/future/base.py]: How long will this wait between calls? I would think that a polling interval should be passed in, in addition to the timeout.

Comment (jonparrott) [inline core/google/cloud/future/base.py]: We trying to stick with the futures interface which just allows you to pass in the timeout. Also, if you use `add_done_callback` you will not have the ability to pass in a timeout or retry interval.

**Comment by jonparrott** [comment]
> I'd prefer if it was
> rows = query_job.result()

That seems reasonable as long as we don't lose data, but wouldn't that require switching to core iterator (#2840)? Could we accept the non-breaking behavior of this PR as-is for now and do another PR to switch the return value of `result` after #2840 is implemented?

> Also, the query job needs to be calling getQueryResults and checking for jobComplete in the response rather than polling the job state. The reason for this is that getQueryResults will wait for the job to complete or timeout (whichever comes first), whereas get Job returns immediately.

Why is that preferable? Doing a long-lived HTTP request is quite different from any of the other behavior in this library. Also, would you be okay with addressing this in a follow-up PR rather than in this one?

**Comment by tswast** [comment]
#2840 is already implemented.

`getQueryResults` is preferable because the job polling method adds up to `polling_interval` milliseconds of latency to get the query results compared to the long-lived request which will return results as soon as they are available.

**Comment by tswast** [comment]
I'm okay with doing the `getQueryResults` in a subsequent PR.

**Comment by jonparrott** [comment]
@tswast Cool, are you good with this PR as-is then?

**Comment by tswast** [comment]
Could we make the change where `QueryJob.result()` returns `QueryResults` in this PR?

**Comment by tswast** [comment]
> Could we make the change where QueryJob.result() returns QueryResults in this PR?

If not, I'll propose we make that as a breaking change when changing the polling method.

**Comment by jonparrott** [comment]
> Could we make the change where QueryJob.result() returns QueryResults in this PR?

Yep. Done. Can you try this out and let me know how broken it is?

This should work:

```python
def async_query(query):
    client = bigquery.Client()
    query_job = client.run_async_query(str(uuid.uuid4()), query)
    query_job.use_legacy_sql = False

    rows = query_job.result().fetch_data(max_results=10)
    for row in rows:
        print(row)
```

**Comment by jonparrott** [comment]
@tswast verified that this seems to work:

```python
>>> from google.cloud import bigquery
>>> import uuid
>>>
>>> client = bigquery.Client()
>>> query = 'SELECT 1'
>>> query_job = client.run_async_query(str(uuid.uuid4()), query)
>>> query_job.use_legacy_sql = False
>>>
>>> rows = query_job.result().fetch_data(max_results=10)
>>> for row in rows:
...     print(row)
...
(1,)
```

File: bigquery/google/cloud/bigquery/job.py [hunk 2]
```diff
@@ -27,6 +31,7 @@
 from google.cloud.bigquery._helpers import UDFResourcesProperty
 from google.cloud.bigquery._helpers import _EnumProperty
 from google.cloud.bigquery._helpers import _TypedProperty
+import google.cloud.future.base
```
Comment (dhermes) [inline bigquery/google/cloud/bigquery/job.py]: This should go up with the other group (the "1st-party-almost" group)

File: bigquery/google/cloud/bigquery/job.py [hunk 3]
```diff
@@ -371,12 +371,83 @@ def cancel(self, client=None):
...
+
+        :rtype: bool
+        :returns: Boolean indicating that the cancel request was sent.
         """
         client = self._require_client(client)
 
         api_response = client._connection.api_request(
             method='POST', path='%s/cancel' % (self.path,))
         self._set_properties(api_response['job'])
+        return True
```
Comment (dhermes) [inline bigquery/google/cloud/bigquery/job.py]: Why?

File: bigquery/google/cloud/bigquery/job.py [hunk 4]
```diff
@@ -371,12 +371,83 @@ def cancel(self, client=None):
...
+        """Set the result or exception from the job if it is complete."""
+        # This must be done in a lock to prevent the polling thread
+        # and main thread from both executing the completion logic
+        # at the same time.
+        with self._completion_lock:
+            # If the operation isn't complete or if the result has already been
+            # set, do not call set_result/set_exception again.
+            # Note: self._result_set is set to True in set_result and
+            # set_exception, in case those methods are invoked directly.
+            if self.state != 'DONE' or self._result_set:
```
Comment (dhermes) [inline bigquery/google/cloud/bigquery/job.py]: Where does `DONE` come from? Make it a constant?

File: bigquery/google/cloud/bigquery/job.py [hunk 5]
```diff
@@ -371,12 +371,83 @@ def cancel(self, client=None):
...
+                return
+
+            if self.error_result is not None:
+                exception = exceptions.GoogleCloudError(
+                    self.error_result, errors=self.errors)
```
Comment (dhermes) [inline bigquery/google/cloud/bigquery/job.py]: What is `error_result`? We have `google.cloud.exceptions.make_exception`

File: bigquery/google/cloud/bigquery/job.py [hunk 6]
```diff
@@ -371,12 +371,83 @@ def cancel(self, client=None):
...
+
+    def done(self):
+        """Refresh the job and checks if it is complete.
+
+        :rtype: bool
+        :returns: True if the job is complete, False otherwise.
+        """
+        # Do not refresh is the state is already done, as the job will not
+        # change once complete.
+        if self.state != 'DONE':
```
Comment (dhermes) [inline bigquery/google/cloud/bigquery/job.py]: OK definitely make `DONE` a constant

File: bigquery/google/cloud/bigquery/job.py [hunk 7]
```diff
@@ -1127,11 +1198,44 @@ def from_api_repr(cls, resource, client):
...
         """Construct a QueryResults instance, bound to this job.
 
         :rtype: :class:`~google.cloud.bigquery.query.QueryResults`
         :returns: results instance
         """
         from google.cloud.bigquery.query import QueryResults
         return QueryResults.from_query_job(self)
+
+    def results(self):
+        """DEPRECATED.
```
Comment (dhermes) [inline bigquery/google/cloud/bigquery/job.py]: Can you add a GitHub Issue and a Google Calendar event marking "when" it should be deprecated?

File: bigquery/google/cloud/bigquery/job.py (hunk 5 above)
Comment (jonparrott) [inline bigquery/google/cloud/bigquery/job.py]: see https://cloud.google.com/bigquery/docs/reference/rest/v2/jobs

It's

```python
debugInfo: str
location: str
message: str
reason: str
```

It does not match HTTP error codes.

Comment (jonparrott) [inline bigquery/google/cloud/bigquery/job.py]: Seems we can map them using https://cloud.google.com/bigquery/troubleshooting-errors

File: bigquery/google/cloud/bigquery/job.py (hunk 3 above)
Comment (jonparrott) [inline bigquery/google/cloud/bigquery/job.py]: Added a comment explaining (It's to satisfy the future interface)

File: bigquery/google/cloud/bigquery/job.py (hunk 4 above)
Comment (jonparrott) [inline bigquery/google/cloud/bigquery/job.py]: Done. (pun intended)

File: bigquery/google/cloud/bigquery/job.py (hunk 6 above)
Comment (jonparrott) [inline bigquery/google/cloud/bigquery/job.py]: (done)

File: bigquery/google/cloud/bigquery/job.py (hunk 7 above)
Comment (jonparrott) [inline bigquery/google/cloud/bigquery/job.py]: #3652

File: bigquery/google/cloud/bigquery/job.py [hunk 8]
```diff
@@ -33,6 +35,43 @@
...
+    'quotaExceeded': http_client.FORBIDDEN,
+    'rateLimitExceeded': http_client.FORBIDDEN,
+    'resourceInUse': http_client.BAD_REQUEST,
+    'resourcesExceeded': http_client.BAD_REQUEST,
+    'responseTooLarge': http_client.FORBIDDEN,
+    'stopped': http_client.OK,
+    'tableUnavailable': http_client.BAD_REQUEST,
+}
+
+_FakeResponse = collections.namedtuple('_FakeResponse', ['status'])
```
Comment (dhermes) [inline bigquery/google/cloud/bigquery/job.py]: You would! :grinning:

File: bigquery/google/cloud/bigquery/job.py [hunk 9]
```diff
@@ -33,6 +35,43 @@
...
+
+
+def _error_result_to_exception(error_result):
+    """"""
```
Comment (dhermes) [inline bigquery/google/cloud/bigquery/job.py]: I really love this artisanal hand-crafted docstring

File: bigquery/google/cloud/bigquery/job.py [hunk 10]
```diff
@@ -33,6 +35,43 @@
...
+    reason = error_result.get('reason')
+    status_code = _ERROR_REASON_TO_EXCEPTION.get(
+        reason, http_client.INTERNAL_SERVER_ERROR)
+    # make_exception expects an httplib2 response object.
+    fake_response = _FakeResponse(status=status_code)
+    return exceptions.make_exception(
+        fake_response, b'', error_info=error_result, use_json=False)
```
Comment (dhermes) [inline bigquery/google/cloud/bigquery/job.py]: `content` should actually be `error_result` (not `b''`)

File: bigquery/google/cloud/bigquery/job.py (hunk 5 above)
Comment (jonparrott) [inline bigquery/google/cloud/bigquery/job.py]: Done.

File: bigquery/google/cloud/bigquery/job.py [hunk 11]
```diff
@@ -447,7 +487,8 @@ def cancelled(self):
         :rtype: bool
         :returns: False
         """
-        return False
+        return (self.error_result is not None
+                and self.error_result.get('reason') == 'stopped')
```
Comment (dhermes) [inline bigquery/google/cloud/bigquery/job.py]: 1. Worth caching this output?
2. Where does `stopped` come from?

File: bigquery/tests/unit/test_job.py [hunk 12]
```diff
@@ -15,9 +15,30 @@
 import copy
 import warnings
 
+from six.moves import http_client
 import unittest
 
 
+class TestErrorResultToException(unittest.TestCase):
```
Comment (dhermes) [inline bigquery/tests/unit/test_job.py]: Preferred name: `Test__error_result_to_exception` (your call)

File: bigquery/google/cloud/bigquery/job.py (hunk 10 above)
Comment (jonparrott) [inline bigquery/google/cloud/bigquery/job.py]: Do I need to pass in the `error_info` in that case?

File: bigquery/google/cloud/bigquery/job.py (hunk 9 above)
Comment (jonparrott) [inline bigquery/google/cloud/bigquery/job.py]: Oops. Fixed in incoming commit.

File: bigquery/google/cloud/bigquery/job.py (hunk 8 above)
Comment (jonparrott) [inline bigquery/google/cloud/bigquery/job.py]: I thought you'd love this.

File: bigquery/google/cloud/bigquery/job.py (hunk 11 above)
Comment (jonparrott) [inline bigquery/google/cloud/bigquery/job.py]: 1. No.
2. It's from the list of reasons above. I meant to add a comment there. Done in incoming commit.

File: bigquery/tests/unit/test_job.py (hunk 12 above)
Comment (jonparrott) [inline bigquery/tests/unit/test_job.py]: Will do.

File: bigquery/google/cloud/bigquery/job.py (hunk 10 above)
Comment (dhermes) [inline bigquery/google/cloud/bigquery/job.py]: `error_info` is [like a "post-script"][1] for the actual error message:

```python
    if error_info is not None:
        message += ' (%s)' % (error_info,)
```

i.e. it's not what you think it is

[1]: https://github.com/GoogleCloudPlatform/google-cloud-python/blob/f8758dfda54c79a7c712e73348fc37a4c349c212/core/google/cloud/exceptions.py#L227-L228

Comment (jonparrott) [inline bigquery/google/cloud/bigquery/job.py]: That doesn't really help me understand what I *should* do here. ISTM that according to the test this is doing what I expect it to do (raise an error with the error result as the message).

I'm happy to change it, but I don't yet understand what I should change.

Comment (dhermes) [inline bigquery/google/cloud/bigquery/job.py]: Can you give me a REPL example of what it does do?

**Comment by jonparrott** [comment]
Thanks @dhermes and @tswast. This is a huge jump in developer experience for bigquery. 🍰

**Pull request merged**